File: app/src/main/java/com/temoa/gankio/ui/activity/FavoritesActivity.java
package com.temoa.gankio.ui.activity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.PopupMenu;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.temoa.gankio.R;
import com.temoa.gankio.bean.NewGankData;
import com.temoa.gankio.db.DBHelper;
import com.temoa.gankio.db.DBTools;
import com.temoa.gankio.ui.adapter.RecyclerAdapter;

import java.util.ArrayList;

public class FavoritesActivity extends AppCompatActivity {

    private RecyclerAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favorites);

        Toolbar toolbar = (Toolbar) findViewById(R.id.fav_activity_toolbar);
        toolbar.setTitle(R.string.collect);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null)
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onBackPressed();
            }
        });

        initRecycler();
        getDataFromDB();
    }

    private void initRecycler() {
        RecyclerView recycler = (RecyclerView) findViewById(R.id.fav_activity_rv);
        recycler.setHasFixedSize(true);
        recycler.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        recycler.setItemAnimator(new DefaultItemAnimator());

        adapter = new RecyclerAdapter(this, recycler, null);
        adapter.isTypeReplaceAuthor(true);
        adapter.addItemClickListener(new RecyclerAdapter.ItemClickListener() {
            @Override
            public void onItemClick(View v, NewGankData.Results data, int position) {
                Intent intent = new Intent();
                intent.setAction("android.intent.action.VIEW");
                intent.setData(Uri.parse(data.getUrl()));
                startActivity(intent);
            }
        });
        adapter.addItemLongClickListener(new RecyclerAdapter.ItemLongClickListener() {
            @Override
            public void onItemLongClick(View v, NewGankData.Results data, int position) {
                createPopupMenu(v, data.getDesc());
            }
        });
        recycler.setAdapter(adapter);
    }

    private void getDataFromDB() {
        DBHelper dbHelper = new DBHelper(this);
        ArrayList<NewGankData.Results> dataList = DBTools.queryAll(dbHelper);
        adapter.setNewData(dataList);
    }

    private void deleteDataInDB(String desc) {
        DBHelper dbHelper = new DBHelper(this);
        DBTools.deleteByDesc(dbHelper, desc);
    }

    private void createPopupMenu(View v, final String desc) {
        final PopupMenu menu = new PopupMenu(this, v);
        menu.getMenuInflater().inflate(R.menu.fav_menu, menu.getMenu());
        menu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                deleteDataInDB(desc);
                menu.dismiss();
                getDataFromDB();
                return true;
            }
        });
        menu.show();
    }
}

File: app/src/main/java/com/temoa/gankio/ui/activity/FavoritesActivity.kt
package com.temoa.gankio.ui.activity

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.lifecycle.Observer
import com.temoa.gankio.R
import com.temoa.gankio.bean.GankData
import com.temoa.gankio.databinding.ActivityFavoritesBinding
import com.temoa.gankio.ui.adapter.GankDataAdapter
import com.temoa.gankio.viewmodel.FavViewModel
import org.koin.androidx.viewmodel.ext.android.viewModel

class FavoritesActivity : AppCompatActivity() {

  private val mFavViewModel: FavViewModel by viewModel()

  private lateinit var mGankDataAdapter: GankDataAdapter

  private lateinit var mViewBinding: ActivityFavoritesBinding

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    mViewBinding = ActivityFavoritesBinding.inflate(layoutInflater)
    setContentView(mViewBinding.root)

    mViewBinding.apply {
      favToolbar.setTitle(R.string.collect)
      setSupportActionBar(favToolbar)
      if (supportActionBar != null) supportActionBar!!.setDisplayHomeAsUpEnabled(true)
      favToolbar.setNavigationOnClickListener { onBackPressed() }

      favRecycler.setHasFixedSize(true)
      mGankDataAdapter = GankDataAdapter(
          onItemClick = { _, data, _ ->
            val intent = Intent()
            intent.action = "android.intent.action.VIEW"
            intent.data = Uri.parse(data.url)
            startActivity(intent)
          },
          onItemLongClick = { view, data, _ ->
            createPopupMenu(view, data)
            true
          }
      ).apply {
        isTypeReplaceAuthor(true)
      }
      favRecycler.adapter = mGankDataAdapter
    }

    mFavViewModel.favLiveData.observe(this, Observer {
      mGankDataAdapter.submitData(lifecycle, it)
    })
  }

  private fun createPopupMenu(v: View, gankData: GankData) {
    val menu = PopupMenu(this, v)
    menu.menuInflater.inflate(R.menu.fav_menu, menu.menu)
    menu.setOnMenuItemClickListener {
      mFavViewModel.delete(gankData)
      true
    }
    menu.show()
  }
}